### PR TITLE
add a windows VM as a jumpbox to Bastion RDP into the Hub network

### PR DESCRIPTION
**Pull request by glennmusa**
# Description

This change proposes adding a Windows VM to allow for Bastion RDP into the Hub network via a new module called `jumpbox`. To demo this, deploy MLZ in your preferred way and initiate a Bastion RDP session to the VM in the hub resource group. To retrieve the login credentials, [add a KeyVault get secret management access policy for your principal](https://docs.microsoft.com/en-us/azure/key-vault/general/assign-access-policy-portal#assign-an-access-policy).

These changes propose:

- Adding a new module called `jumpbox`
  - creates a KeyVault resource
  - creates two secret values for jumpbox username and jumpbox password to the KeyVault
  - calls the `windows-virtual-machine` module to create a VM and supply these credentials
- Adds a `jumpbox-subnet` subnet module to `saca-hub` to set up traffic flow logs and NSG rules for SSH and RDP
- Adds `mlz_objectid` as a variable to reference the MLZ Service Principal's AAD object ID for assigning KeyVault policies

## Issue reference

The issue this PR will close: #187, close: #190 

## Checklist

Please make sure you've  completed the relevant tasks for this PR, out of the following list:

* [x] Code compiles or validates correctly
* [x] BASH scripts have been validated using `shellcheck`
* [x] All tests pass (manual and automated)
* [x] The documentation is updated to cover any new or changed features
* [x] Markdown files have been linted using the recommended linter. (See `.vscode/extensions.json`.)
* [x] Relevant issues are linked to this PR

### CODE REVIEW

File: src/core/saca-hub/variables.tf [hunk 1]
```diff
@@ -140,3 +144,112 @@ variable "bastion_ipconfig_name" {
...
+      }
+    }
+
+    routetable_name = "mlzDemoJumpboxSubnetRt"
+  }
+}
+
+variable "jumpbox_keyvault_name" {
+  description = "The name of the jumpbox virtual machine keyvault"
+  default     = "mlzDemoJumpboxVmKv"
```
Comment (brooke-hamilton) [inline src/core/saca-hub/variables.tf:217]: My deployment got this error, which I think is related to this value.

```
Error: creating Vault: (Name "mlzDemoJumpboxVmKv" / Resource Group "rg-saca-jbh003"): keyvault.VaultsClient#CreateOrUpdate: Failure sending request: StatusCode=0 -- Original Error: autorest/azure: Service returned an error. Status=<nil> Code="VaultAlreadyExists" Message="The vault name 'mlzDemoJumpboxVmKv' is already in use. Vault names are globaly unique so it is possible that the name is already taken. If you are sure that the vault name was not taken then it is possible that a vault with the same name was recently deleted but not purged after being placed in a recoverable state. If the vault is in a recoverable state then the vault will need to be purged before reusing the name. For more information on soft delete and purging a vault follow this link https://go.microsoft.com/fwlink/?linkid=2147740."
```

File: src/core/saca-hub/saca-hub.front.json [hunk 2]
```diff
@@ -177,6 +177,85 @@
...
+				"varname": "jumpbox_subnet.routetable_name",
+				"type": "text",
+				"default_val": "mlzDemoJumpboxSubnetRt",
+				"description": "The name of the jumpbox subnet route table",
+				"options": []
+			},
+			{
+				"varname": "jumpbox_keyvault_name",
+				"type": "text",
+				"default_val": "mlzDemoJumpboxVmKv",
```
Comment (brooke-hamilton) [inline src/core/saca-hub/saca-hub.front.json:214]: The error I got when deploying could also be related to this default.

**Review by brooke-hamilton — CHANGES_REQUESTED** [review]

File: src/core/saca-hub/variables.tf (hunk 1 above)
Comment (glennmusa) [reply src/core/saca-hub/variables.tf:217]: Ah. KeyVaults are a global resource. My mistake. Will generate a name.

File: src/core/saca-hub/saca-hub.front.json (hunk 2 above)
Comment (glennmusa) [reply src/core/saca-hub/saca-hub.front.json:214]: Will generate a name

File: src/modules/jumpbox/main.tf [hunk 3]
```diff
@@ -0,0 +1,71 @@
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT License.
+
+resource "random_id" "jumpbox-keyvault" {
+  byte_length = 12
+}
+
+resource "azurerm_key_vault" "jumpbox-keyvault" {
+  name                       = format("%.24s", lower(replace("${var.keyvault_name}${random_id.jumpbox-keyvault.id}", "/[[:^alnum:]]/", "")))
```
Comment (glennmusa) [inline src/modules/jumpbox/main.tf:9]: @brooke-hamilton now appending a random alphanumeric string to the passed in keyvault name

**Review by brooke-hamilton — CHANGES_REQUESTED** [review]
I'm getting this error when deploying in this way:
```
src/scripts/deploy.sh -s <subscriptionid> -z <deployment name>
```

```
Error: A resource with the ID "/subscriptions/<sub id>/resourcegroups/rg-t0-jbh006/providers/microsoft.network/virtualnetworks/vn-t0-jbh006/providers/microsoft.insights/diagnosticSettings/vn-t0-jbh006-vn-diagnostics" already exists - to be managed via Terraform this resource needs to be imported into the State. Please see the resource documentation for "azurerm_monitor_diagnostic_setting" for more information.

  on ../../modules/virtual-network/main.tf line 30, in resource "azurerm_monitor_diagnostic_setting" "vnet":
  30: resource "azurerm_monitor_diagnostic_setting" "vnet" {


ERROR: failed to apply tier-0 (1/5). Trying some manual clean-up and Terraform destroy...
Initializing modules...
```

**Comment by glennmusa** [comment]
> I'm getting this error when deploying in this way:
> 
> ```
> src/scripts/deploy.sh -s <subscriptionid> -z <deployment name>
> ```
> 
> ```
> Error: A resource with the ID "/subscriptions/2b01c398-5868-496e-a84f-503ea9eac23a/resourcegroups/rg-t0-jbh006/providers/microsoft.network/virtualnetworks/vn-t0-jbh006/providers/microsoft.insights/diagnosticSettings/vn-t0-jbh006-vn-diagnostics" already exists - to be managed via Terraform this resource needs to be imported into the State. Please see the resource documentation for "azurerm_monitor_diagnostic_setting" for more information.
> 
>   on ../../modules/virtual-network/main.tf line 30, in resource "azurerm_monitor_diagnostic_setting" "vnet":
>   30: resource "azurerm_monitor_diagnostic_setting" "vnet" {
> 
> 
> ERROR: failed to apply tier-0 (1/5). Trying some manual clean-up and Terraform destroy...
> Initializing modules...
> ```

`A resource with the ID "/subscriptions/2b01c398-5868-496e-a84f-503ea9eac23a/resourcegroups/rg-t0-jbh006/providers/microsoft.network/virtualnetworks/vn-t0-jbh006/providers/microsoft.insights/diagnosticSettings/vn-t0-jbh006-vn-diagnostics" already exists - to be managed via Terraform this resource needs to be imported into the State.` is an error unrelated to these changes, but is encountered by diagnostic settings that are in an unknown state. Typically you'd see this when a resource is created by Terraform, but deleted manually. The solution is to recreate the resources at the ID signature and delete them using Terraform: https://github.com/terraform-providers/terraform-provider-azurerm/issues/8109

**Comment by brooke-hamilton** [comment]
> > I'm getting this error when deploying in this way:
> > ```
> > src/scripts/deploy.sh -s <subscriptionid> -z <deployment name>
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > ```
> > Error: A resource with the ID "/subscriptions/<sub id>/resourcegroups/rg-t0-jbh006/providers/microsoft.network/virtualnetworks/vn-t0-jbh006/providers/microsoft.insights/diagnosticSettings/vn-t0-jbh006-vn-diagnostics" already exists - to be managed via Terraform this resource needs to be imported into the State. Please see the resource documentation for "azurerm_monitor_diagnostic_setting" for more information.
> > 
> >   on ../../modules/virtual-network/main.tf line 30, in resource "azurerm_monitor_diagnostic_setting" "vnet":
> >   30: resource "azurerm_monitor_diagnostic_setting" "vnet" {
> > 
> > 
> > ERROR: failed to apply tier-0 (1/5). Trying some manual clean-up and Terraform destroy...
> > Initializing modules...
> > ```
> 
> `A resource with the ID "/subscriptions/<sub id>/resourcegroups/rg-t0-jbh006/providers/microsoft.network/virtualnetworks/vn-t0-jbh006/providers/microsoft.insights/diagnosticSettings/vn-t0-jbh006-vn-diagnostics" already exists - to be managed via Terraform this resource needs to be imported into the State.` is an error unrelated to these changes, but is encountered by diagnostic settings that are in an unknown state. Typically you'd see this when a resource is created by Terraform, but deleted manually. The solution is to recreate the resources at the ID signature and destroy them using Terraform: [terraform-providers/terraform-provider-azurerm#8109](https://github.com/terraform-providers/terraform-provider-azurerm/issues/8109)

Thanks for the help on this - it's an issue that happens if I create the resource using terraform and then delete it using the Azure CLI or Azure portal, and then recreate the resource using the same name. If I delete the resource using terraform then the issue does not happen (or if I use a unique name).

**Review by brooke-hamilton — APPROVED** [review]
I was able to deploy and log into the VM via Bastion. 🥇

**Pull request merged**